cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+      this.programId = programId;
+      this.programType = programType;
+      this.streamSizeSchedule = streamSizeSchedule;
+    }
+
+    public void startSchedule(long baseSize, long baseTs, boolean active, boolean persist) {
+      LOG.debug("Starting schedule {} with baseSize {}, baseTs {}, active {}. Should be persisted: {}",
+                streamSizeSchedule.getName(), baseSize, baseTs, active, persist);
+      this.baseSize = baseSize;
+      this.baseTs = baseTs;
+      this.running = active;
+    }
+
+    public boolean isRunning() {
+      return running;
+    }
+
+    public void received(StreamSizeNotification notification) {
+      if (!running) {
+        return;
+      }
+      long pastRunSize;
+      long pastRunTs;
+      synchronized (this) {
+        if (notification.getSize() < baseSize) {
+          // This can happen when a stream is truncated: the baseSize is still the old size,
+          // but we receive notification with way less data
+          baseSize = notification.getSize();
+          baseTs = notification.getTimestamp();
+          return;
+        }
+        if (notification.getSize() < baseSize + toBytes(streamSizeSchedule.getDataTriggerMB())) {
+          return;
+        }
+
+        // Update the baseSize as soon as possible to avoid races
+        pastRunSize = baseSize;
+        pastRunTs = baseTs;
+        baseSize = notification.getSize();
+        baseTs = notification.getTimestamp();
+        LOG.debug("Base size and ts updated to {}, {} for streamSizeSchedule {}",
+                  baseSize, baseTs, streamSizeSchedule);
+      }
+
+      Arguments args = new BasicArguments(ImmutableMap.of(
+        ProgramOptionConstants.SCHEDULE_NAME, streamSizeSchedule.getName(),
+        ProgramOptionConstants.LOGICAL_START_TIME, Long.toString(baseTs),
+        ProgramOptionConstants.RUN_DATA_SIZE, Long.toString(baseSize),
+        ProgramOptionConstants.PAST_RUN_LOGICAL_START_TIME, Long.toString(pastRunTs),
+        ProgramOptionConstants.PAST_RUN_DATA_SIZE, Long.toString(pastRunSize)
+      ));
+
+      while (true) {
+        ScheduleTaskRunner taskRunner = new ScheduleTaskRunner(getStore(), programRuntimeService, preferencesStore);
+        try {
+          LOG.info("About to start streamSizeSchedule {}", streamSizeSchedule);
+          taskRunner.run(programId, ProgramType.valueOf(programType.name()), args);
+          break;
+        } catch (TaskExecutionException e) {
+          LOG.error("Execution exception while running streamSizeSchedule {}", streamSizeSchedule.getName(), e);
+          if (e.isRefireImmediately()) {
+            LOG.info("Retrying execution for streamSizeSchedule {}", streamSizeSchedule.getName());
+          } else {
+            break;
+          }
+        }
+      }
+    }
+
+    /**
+     * @return true if we successfully suspended the schedule, false if it was already suspended
+     */
+    public boolean suspend() {
```

Comment:
should this be synchronized or use an atomic? If two threads call this concurrently, they may both return true. I believe that it may get called concurrently.